File: ios/PowerPlayHockeyCore/Sources/PowerPlayHockeyCore/Model/TournamentData.swift
import Foundation

struct TournamentMatch: Codable, Equatable {
    let team1: Int
    let team2: Int
    var score1: Int = -1
    var score2: Int = -1
    var winner: Int = -1

    var isPlayed: Bool { winner != -1 }

    init(team1: Int, team2: Int, score1: Int = -1, score2: Int = -1, winner: Int = -1) {
        self.team1 = team1
        self.team2 = team2
        self.score1 = score1
        self.score2 = score2
        self.winner = winner
    }

    enum CodingKeys: String, CodingKey {
        case team1 = "t1"
        case team2 = "t2"
        case score1 = "s1"
        case score2 = "s2"
        case winner = "w"
    }
}

final class TournamentState: Codable {
    let userTeam: Int
    let league: String
    var currentRound: Int // 0 = QF, 1 = SF, 2 = Final, 3 = Complete
    var qf: [TournamentMatch]
    var sf: [TournamentMatch]
    var finals: TournamentMatch?
    var userEliminated: Bool
    var trophyWon: Bool

    init(
        userTeam: Int,
        league: String,
        currentRound: Int = 0,
        qf: [TournamentMatch] = [],
        sf: [TournamentMatch] = [],
        finals: TournamentMatch? = nil,
        userEliminated: Bool = false,
        trophyWon: Bool = false
    ) {
        self.userTeam = userTeam
        self.league = league
        self.currentRound = currentRound
        self.qf = qf
        self.sf = sf
        self.finals = finals
        self.userEliminated = userEliminated
        self.trophyWon = trophyWon
    }

    static func createNew(userTeamIdx: Int, rng: inout SeededRandom) -> TournamentState {
        let userTeam = TeamInfo.byIndex(userTeamIdx)
        let league = userTeam.league
        var allInLeague = TeamInfo.ALL.indices.filter { TeamInfo.byIndex($0).league == league && $0 != userTeamIdx }
        
        // Fisher-Yates shuffle with SeededRandom
        for i in (1..<allInLeague.count).reversed() {
            let j = Int(rng.nextFloat() * Float(i + 1))
            if i != j { allInLeague.swapAt(i, j) }
        }
        let opponents = Array(allInLeague.prefix(7))

        let state = TournamentState(userTeam: userTeamIdx, league: league)
        if opponents.count >= 7 {
            state.qf.append(TournamentMatch(team1: userTeamIdx, team2: opponents[0]))
            state.qf.append(TournamentMatch(team1: opponents[1], team2: opponents[2]))
            state.qf.append(TournamentMatch(team1: opponents[3], team2: opponents[4]))
            state.qf.append(TournamentMatch(team1: opponents[5], team2: opponents[6]))
        }
        return state
    }

    static func simulateAiMatch(match: inout TournamentMatch, rng: inout SeededRandom) {
        var s1 = Int(rng.nextFloat() * 4) + 1
        var s2 = Int(rng.nextFloat() * 4) + 1
        if s1 == s2 {
            if rng.nextBool() { s1 += 1 } else { s2 += 1 }
        }
        match.score1 = s1
        match.score2 = s2
        match.winner = s1 > s2 ? match.team1 : match.team2
    }
}

File: app/src/main/java/com/tablehockey/game/model/TournamentData.kt
package com.tablehockey.game.model

import org.json.JSONArray
import org.json.JSONObject
import java.io.Serializable
import kotlin.random.Random

data class TournamentMatch(
    val team1: Int,
    val team2: Int,
    var score1: Int = -1,
    var score2: Int = -1,
    var winner: Int = -1
) : Serializable {
    val isPlayed: Boolean get() = winner != -1

    fun toJson(): JSONObject = JSONObject().apply {
        put("t1", team1)
        put("t2", team2)
        put("s1", score1)
        put("s2", score2)
        put("w", winner)
    }

    companion object {
        fun fromJson(json: JSONObject): TournamentMatch {
            return TournamentMatch(
                team1 = json.getInt("t1"),
                team2 = json.getInt("t2"),
                score1 = json.optInt("s1", -1),
                score2 = json.optInt("s2", -1),
                winner = json.optInt("w", -1)
            )
        }
    }
}

data class TournamentState(
    val userTeam: Int,
    val league: String,
    var currentRound: Int = 0, // 0 = QF, 1 = SF, 2 = Final, 3 = Complete
    val qf: MutableList<TournamentMatch> = mutableListOf(),
    val sf: MutableList<TournamentMatch> = mutableListOf(),
    var finals: TournamentMatch? = null,
    var userEliminated: Boolean = false,
    var trophyWon: Boolean = false
) : Serializable {

    fun toJson(): String {
        val root = JSONObject()
        root.put("userTeam", userTeam)
        root.put("league", league)
        root.put("currentRound", currentRound)
        root.put("userEliminated", userEliminated)
        root.put("trophyWon", trophyWon)

        val qfArray = JSONArray()
        for (m in qf) qfArray.put(m.toJson())
        root.put("qf", qfArray)

        val sfArray = JSONArray()
        for (m in sf) sfArray.put(m.toJson())
        root.put("sf", sfArray)

        if (finals != null) {
            root.put("finals", finals!!.toJson())
        }
        return root.toString()
    }

    companion object {
        fun fromJson(str: String): TournamentState? {
            return try {
                val root = JSONObject(str)
                val state = TournamentState(
                    userTeam = root.getInt("userTeam"),
                    league = root.getString("league"),
                    currentRound = root.getInt("currentRound"),
                    userEliminated = root.optBoolean("userEliminated", false),
                    trophyWon = root.optBoolean("trophyWon", false)
                )
                val qfArray = root.getJSONArray("qf")
                for (i in 0 until qfArray.length()) {
                    state.qf.add(TournamentMatch.fromJson(qfArray.getJSONObject(i)))
                }
                val sfArray = root.getJSONArray("sf")
                for (i in 0 until sfArray.length()) {
                    state.sf.add(TournamentMatch.fromJson(sfArray.getJSONObject(i)))
                }
                if (root.has("finals")) {
                    state.finals = TournamentMatch.fromJson(root.getJSONObject("finals"))
                }
                state
            } catch (e: Exception) {
                null
            }
        }

        fun createNew(userTeamIdx: Int, rng: Random = Random): TournamentState {
            val userTeam = TeamInfo.byIndex(userTeamIdx)
            val league = userTeam.league
            val allInLeague = TeamInfo.ALL.indices.filter { TeamInfo.byIndex(it).league == league && it != userTeamIdx }.shuffled(rng)
            val opponents = allInLeague.take(7)

            val state = TournamentState(userTeam = userTeamIdx, league = league)

            state.qf.add(TournamentMatch(userTeamIdx, opponents[0]))
            state.qf.add(TournamentMatch(opponents[1], opponents[2]))
            state.qf.add(TournamentMatch(opponents[3], opponents[4]))
            state.qf.add(TournamentMatch(opponents[5], opponents[6]))

            return state
        }

        fun simulateAiMatch(match: TournamentMatch, rng: Random = Random) {
            var s1 = rng.nextInt(1, 5)
            var s2 = rng.nextInt(1, 5)
            if (s1 == s2) {
                if (rng.nextBoolean()) s1++ else s2++
            }
            match.score1 = s1
            match.score2 = s2
            match.winner = if (s1 > s2) match.team1 else match.team2
        }
    }
}
